components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractLocalApplicationAuthenticator.java

```diff
@@ -0,0 +1,289 @@
+        implements ApplicationAuthenticator {
+
+    private static final long serialVersionUID = -4406878411547612129L;
+    private static final Log log = LogFactory.getLog(AbstractLocalApplicationAuthenticator.class);
+
+    @Override
+    public AuthenticatorFlowStatus process(HttpServletRequest request, HttpServletResponse response,
+                                           AuthenticationContext context) throws AuthenticationFailedException,
+            LogoutFailedException {
+
+        // if an authentication flow
+        if (!context.isLogoutRequest()) {
+            if (!canHandle(request)
+                    || Boolean.TRUE.equals(request.getAttribute(FrameworkConstants.REQ_ATTR_HANDLED))) {
+                context.setRetrying(false);
+                return initiateAuthenticationFlow(request, response, context);
+            } else {
+                try {
+                    fireEvent(context, IdentityEventConstants.Event.PRE_AUTHENTICATION, false);
+                    processAuthenticationResponse(request, response, context);
+                    if (this instanceof LocalApplicationAuthenticator && !context.getSequenceConfig()
+                            .getApplicationConfig().isSaaSApp()) {
+                        processWithNonSaaSApp(context);
+                    }
+                    request.setAttribute(FrameworkConstants.REQ_ATTR_HANDLED, true);
+                    context.setProperty(FrameworkConstants.LAST_FAILED_AUTHENTICATOR, null);
+                    fireEvent(context, IdentityEventConstants.Event.POST_AUTHENTICATION, true);
+                    return AuthenticatorFlowStatus.SUCCESS_COMPLETED;
+                } catch (AuthenticationFailedException e) {
+                    if (checkUserAccountStatus(response, context)) {
+                        return AuthenticatorFlowStatus.INCOMPLETE;
+                    }
+                    fireEvent(context, IdentityEventConstants.Event.POST_AUTHENTICATION, false);
+                    request.setAttribute(FrameworkConstants.REQ_ATTR_HANDLED, true);
+                    // Decide whether we need to redirect to the login page to retry authentication.
+                    return redirectToRetryAuthentication(request, response, context, e);
+                }
+            }
+            // if a logout flow
+        } else {
+            return processLogoutFlow(request, response, context);
+        }
+    }
+
+    /**
+     * To decide whether need to redirect the user to login page to retry authentication.
+     *
+     * @param request  the httpServletRequest
+     * @param response the httpServletResponse
+     * @param context  the authentication context
+     * @param e        the authentication failed exception
+     * @return authentication flow status
+     * @throws AuthenticationFailedException the exception in the authentication flow
+     */
+    protected AuthenticatorFlowStatus redirectToRetryAuthentication(HttpServletRequest request,
```

Comment:
this method name doesn't seem to reflect what it's doing. May be handleRetryOnFailure()